addons/hr_timesheet/hr_timesheet.py

```diff
@@ -103,129 +77,118 @@ def on_change_unit_amount(self, cr, uid, id, prod_id, unit_amount, company_id, u
-        }}
-
-class account_analytic_account(osv.osv):
+            raise exceptions.ValidationError("This user or employee is not associated to a valid Product")
+    @api.model
+    def _get_employee_unit(self, user_id=None):
+            emp = self.env['hr.employee'].search([('user_id', '=', self.user_id.id or user_id or self.env.uid)])
+            if emp and emp[0].product_id.uom_id.id:
+                return emp[0].product_id.uom_id.id
+            else:
+                raise exceptions.ValidationError("This user or employee is not associated to a valid Product Amount Type")    
+
+    @api.model
+    def _get_general_account(self, user_id=None):
+        emp = self.env['hr.employee'].search([('user_id', '=', self.user_id.id or user_id or self.env.uid)])
+        if emp and emp[0].product_id.categ_id.property_account_expense_categ.id:
+            return emp[0].product_id.categ_id.property_account_expense_categ.id
+        else:
+            raise exceptions.ValidationError("This user or employee is not associated to a valid Product Financial Account")
 
+    @api.model
+    def _get_analytic_journal(self, user_id=None):
+        emp = self.env['hr.employee'].search([('user_id','=',self.user_id.id or user_id or self.env.uid)])
+        if emp and emp[0].journal_id.id:
+            return emp[0].journal_id.id
+        else:
+            raise exceptions.ValidationError("This user or employee is not associated to a valid Journal ID")
+
+    is_timesheet = fields.Boolean()
+   
+    @api.constrains('user_id')
+    def check_user_id(self):
+        if self.is_timesheet:
+            emp = self.env['hr.employee'].search([('user_id','=',self.user_id.id or self.env.uid)])
+            if not emp:
+                raise exceptions.ValidationError("There is no employee defined for user " + self.user_id.name)
+            else:
+                if not emp[0].journal_id.id:
+                    raise exceptions.ValidationError("The employee " + emp[0].name + " is not associated to a valid Analytic Journal. Please define one for him or select another user.")
+                elif not emp[0].product_id:
+                    raise exceptions.ValidationError("The employee " + emp[0].name + " is not associated to a valid Product. Please define one for him or select another user.")
+                elif not emp[0].product_id.uom_id.id:
+                    raise exceptions.ValidationError("The employee " + emp[0].name + " is not associated to a valid Product Unit of Measure. Please define one for him or select another user.")
+                elif not emp[0].product_id.categ_id.property_account_expense_categ.id:
+                    raise exceptions.ValidationError("The employee " + emp[0].name + " is not associated to a valid Financial Account. Please define one for him or select another user.")
+
+    @api.onchange('user_id')
+    def V8_on_change_user_id(self):
+        if self.is_timesheet:
```

Comment:
To be complete, I would remove this condition and make a generic wrapper, unless there is something else intended in the code.
onchange_values = self.on_change_user_id(..).get('value', {})
for key, value in onchange_values.iteritems():
  setattr(self, key, value)